src/main/webapp/meal-page-script.js

```diff
@@ -39,21 +42,79 @@ function createElementByTag(text, tag) {
   return element;
 }
 
-/** Creates a map and adds it to the page. */
+let map;
+
 function createMap() {
-  const userLocation = new google.maps.LatLng(55.746514, 37.627022);
+  let userLocation = new google.maps.LatLng(55.746514, 37.627022);
   const mapOptions = {
-    zoom: 16,
-    center: userLocation
+    zoom: 14,
+    center: userLocation,
   };
-  const map = new google.maps.Map(document.getElementById("map"), mapOptions);
-  // TODO(grenlayk): implement real geolocation here
-  // TODO(grenlayk): change "You are here" marker on custom one
+  map = new google.maps.Map(document.getElementById("map"), mapOptions);
   const locationMarker = new google.maps.Marker({
     position: userLocation,
     map,
-    title: "You are here",
+    title: "You are here (probably)",
+    label: "U",
     animation: google.maps.Animation.DROP,
   });
+
+  
+  if (navigator.geolocation) {
```

Comment:
Oh, so roughly speaking, we have a function with a rather unfriendly API (`getCurrentPosition`). One common way to work with such functions is to wrap them in another custom function with a more friendly API.

This callback-style code was common before introducing Promises [some more details here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises). It often leads to the problem of having to pass the same callback multiple times and being quite hard to follow (aka the callback hell).

The good news is that all of those callback-style functions can be rewritten in terms of promises. The `Promise` object has a constructor of
```
const p = new Promise((resolve, reject) => {
  // some asynchronous code here. The key bit is that `p` will be a Promise, so you can hide the oldschool api here and never think about it again
});
```

If you look at the constructor signature, it's:
1. a function (like every constructor)
2. that takes a function as an argument (the argument passed to `Promise(...)` has a `=>`, therefore it's a function
3. which takes two functions as its argument (both `resolve` and `reject` are functions)
I understand that this may seem a bit weird, so I'm not gonna bore you with the details, but this kinda functional programming (here functional meaning - treating functions just like any other objects, e.g. passing them around as arguments, returning them as values, etc.) is quite common in JS, but can be minimized if someone doesn't feel comfortable working with them.
Btw. those functions that operate on functions are often called [Higher Order Functions](https://en.wikipedia.org/wiki/Higher-order_function) (or HOFs for short).

Either way, you can hide the oldschool geolocation object behind a modern API with this magic function:
```
function getCurrentPositionPromise() {
  return new Promise((resolve, reject) => {
    if (!navigator.geolocation) {
      reject("NO_GEOLOCATION");  // Rejections can be read with a `Promise.catch` method.
    } else {
      const onSuccess = (position) => resolve(position);  // This value will be available on the first `Promise.then` call.
      const onError = () => reject("GET_POSITION_FAILED");
      navigator.geolocation.getCurrentPosition(onSuccess, onError);
    }
  });
}
```
and use it like any other promise, e.g.
```
getCurrentPositionPromise().then(pos => console.log("Current position is:", pos)).catch(err => console.log("this bit will only be called if the promise was rejected", err)).then(() => console.log("another callback"));
```

If you want to experiment with it, you can just paste the function into your browser's console (F12).

Reply:
Thanks!